portal-impl/src/com/liferay/portlet/usersadmin/reindexer/UserReindexerImpl.java

```diff
@@ -0,0 +1,223 @@
+		}
+
+		if (ProxyModeThreadLocal.isForceSync()) {
+			_reindex(newUserIndexerRequests);
+		}
+		else {
+			_queuedIndexerRequestExecutorService.submit(
+				() -> _syncReindex(newUserIndexerRequests));
+		}
+	}
+
+	@Override
+	public void reindex(final User user) throws SearchException {
+		reindex(user.getUserId());
+	}
+
+	private void _reindex(List<UserIndexerRequest> userIndexerRequests) {
+		int i = 0;
+
+		for (UserIndexerRequest userIndexerRequest : userIndexerRequests) {
+			i++;
+
+			if (userIndexerRequest.isExecuting()) {
+				if (_log.isDebugEnabled()) {
+					_log.debug(
+						StringBundler.concat(
+							"Currently executing indexer request ", i,
+							" in another thread: ", userIndexerRequest));
+				}
+
+				continue;
+			}
+
+			synchronized (userIndexerRequest) {
+				if (userIndexerRequest.isIgnore()) {
+					if (_log.isDebugEnabled()) {
+						_log.debug(
+							StringBundler.concat(
+								"Ignoring indexer request ", i,
+								" due to flag set in another thread: ",
+								userIndexerRequest));
+					}
+
+					continue;
+				}
+
+				if (userIndexerRequest.isExecuted()) {
+					if (_log.isDebugEnabled()) {
+						_log.debug(
+							StringBundler.concat(
+								"Already executed indexer request", i,
+								"in another thread: ", userIndexerRequest));
+					}
+
+					continue;
+				}
+
+				if (_log.isDebugEnabled()) {
+					_log.debug(
+						StringBundler.concat(
+							"Executing indexer request ", i, ": ",
+							userIndexerRequest));
+				}
+
+				try {
+					userIndexerRequest.execute();
+				}
+				finally {
+					_queuedIndexerRequests.remove(userIndexerRequest);
+				}
+			}
+		}
+	}
+
+	private void _syncReindex(List<UserIndexerRequest> userIndexerRequests) {
+		try (SafeClosable safeClosable =
+				ProxyModeThreadLocal.setWithSafeClosable(true)) {
+
+			_reindex(userIndexerRequests);
+		}
+	}
+
+	private static final Log _log = LogFactoryUtil.getLog(
+		UserReindexerImpl.class);
+
+	private final ExecutorService _queuedIndexerRequestExecutorService =
+		Executors.newSingleThreadExecutor();
+	private final ConcurrentMap<UserIndexerRequest, UserIndexerRequest>
+		_queuedIndexerRequests = new ConcurrentHashMap<>();
+
+	private class UserIndexerRequest {
+
+		public UserIndexerRequest(long userId, Indexer<User> indexer) {
+			_userId = userId;
+			_indexer = indexer;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if (!(o instanceof UserIndexerRequest)) {
+				return false;
+			}
+
+			UserIndexerRequest other = (UserIndexerRequest)o;
+
+			return _userId.equals(other._userId);
+		}
+
+		public void execute() {
+			if (_ignore) {
+				return;
+			}
+
+			try {
+				_executing = true;
+
+				_indexer.reindex(User.class.getName(), _userId);
+
+				_executed = true;
+			}
+			catch (SearchException se) {
+				_log.error(se, se);
+			}
+			finally {
+				_executing = false;
+			}
+		}
+
+		@Override
+		public int hashCode() {
+			return _userId.hashCode();
+		}
+
+		public void ignore() {
```

Comment:
@brandizzi @holatuwol no one ever calls this ignore() method, so the _ignore is always false. All the readers to this method is dead code.